src/amplify-migration/jsDomUtils.ts

```diff
@@ -0,0 +1,186 @@
+  setOfAllDocumentsPath,
+  currentRepoName,
+}: {
+  dom: JSDOM;
+  tagAttribute: TagAttribute<"a" | "img">;
+  changedPermalinks: { [oldPermalink: string]: string };
+  fileContent:string
+  setOfAllDocumentsPath: Set<string>;
+  normalisedUrls: Set<string>;
+  currentRepoName: string;
+}): Promise<{
+  fileContent: string;
+  dom: JSDOM;
+  changedPermalinks: { [oldPermalink: string]: string };
+}> {
+
+  for (const tag of dom.window.document.querySelectorAll(tagName)) {
+    // replace permalinks with lowercase and in changedPermalinks
+    if (attribute === "href") {
+      /**
+       * Ideally this code should be const a = tag as HTMLAnchorElement;
+       * However, HTMLAnchorElement is a built-in interface in the browser's DOM API,
+       * which means that it's only available in a browser environment.
+       * In development, hovering over tag will show the inferred type as HTMLAnchorElement
+       */
+      const a = tag as any;
+
+      let rawPermalink = getRawPermalink(a[attribute]);
+
+      if (changedPermalinks[rawPermalink]) {
+        a[attribute] = a.href.replace(
+          rawPermalink,
+          changedPermalinks[rawPermalink]
+        );
+        fileContent = fileContent.replace(
+          rawPermalink,
+          changedPermalinks[rawPermalink]
+        );
+      }
+
+      const { fileContent: href } =
+        await updateFilesUploadsPath(
+          a.href,
+          setOfAllDocumentsPath,
+          currentRepoName
+        );
+      if (a.href !== href) {
+        fileContent.replace(a.href, href)
+      }
+
+    } else if (attribute === "src") {
+      /**
+       * Ideally this code should be const img = tag as HTMLImageElement;
+       * However, HTMLImageElement is a built-in interface in the browser's DOM API,
+       * which means that it's only available in a browser environment.
+       * In development, hovering over tag will show the inferred type as HTMLImageElement
+       */
+      const img = tag as any;
+
+      let rawPermalink = getRawPermalink(img[attribute]);
+
+      if (changedPermalinks[rawPermalink]) {
+        img[attribute] = img.src.replace(
+          rawPermalink,
+          changedPermalinks[rawPermalink]
+        );
+      }
+
+      const { fileContent:src } =
+        await updateFilesUploadsPath(
+          img.src,
+          setOfAllDocumentsPath,
+          currentRepoName
+        );
+      if (img.src !== src) {
+        fileContent.replace(img.src, src)
+      }
+    }
+  }
+  return { fileContent, dom, changedPermalinks };
+}
+/**
+ * Requirements:
+ * convert `/some/path` to `some/path`
+ * convert `some/path/` to `some/path`
+ * above two variants with `permalink: ` prefix
+ * above two variants with `url: ` prefix
+ * @param permalink original permalink
+ * @returns raw permalinks without leading/trailing slash
+ */
+
+export function getRawPermalink(permalink: string) {
+  let trimmedPermalink = permalink.trim();
+  if (trimmedPermalink.startsWith(`permalink: `)) {
```

Comment:
Is it always `permalink: ` or can be `permalink:` too without the space?

Reply:
this was a good point I didnt consider! Upon further digging, it seems that `permalink:some/path` (ie without the space) is not __valid__ [YAML syntax](https://stackoverflow.com/questions/42124227/why-does-the-yaml-spec-mandate-a-space-after-the-colon). 

As such, I am inclined to fix iff we are sure that it is valid syntax, so that we keep the scope of the script to solve the problem of: "fix only what is needed to be fixed for the migration to be considered a success"

Reply:
okay!